### PR TITLE
New Git Manager options and changes

### PR DESCRIPTION
**Pull request by lainets**
# Description

**What?**

Adds new Git Manager options.

**Why?**

Git Manager doesn't host the exercises anymore, so we need a way to specify where the exercises are. unprotected_paths option is also required on some courses.

# Testing

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [ ] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [X] Manual testing.

Exercises and chapters seem to work.

**Did you test the changes in**

- [ ] Chrome
- [ ] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Programming style

- [X] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [X] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

**Review by Mankro — COMMENTED** [review]
Good! I still need to test this myself before accepting.

File: README.md [hunk 1]
```diff
@@ -271,7 +287,17 @@ Some fields require a value from specific choices (see also
 
 ## List of directives and examples
 
-### 1. Graded questionnaire
+### 1. Common options for questionnaires and submittable exercises
+
+* `url`: this is the URL where the exercise can be found. Supports formatting
+where `{key}` is replaced with the exercise key. E.g. "https://grader/{key}"
+* `configure-url`: this is the URL with which the exercise is configured.
```
[the remark this answers is not in the document]
Comment (lainets) [reply README.md:295]: That is the URL gitmanager uses to configure the exercise on a grader. Would "this is the URL that is used to configure the exercises on a grader. It is dependent on the grading service used, and is <grader-host>/configure for MOOC-grader." be sufficient?

File: directives/abstract_exercise.py [hunk 2]
```diff
@@ -51,3 +52,66 @@ def set_assistant_permissions(self, data):
...
+        if "url" not in data and env.config.default_exercise_url:
+            data["url"] = env.config.default_exercise_url
+
+        data['url'] = data['url'].format(key=name)
+
+    def set_configure(self, data, exercise_url, files):
+        if 'no-configure' in self.options:
+            return
+
+        if "configure-url" in data:
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/abstract_exercise.py:94]: configure-url isn't actually supposed to be written to the YAML file: that line exists so that configure-url can be overridden using the category and the override config value. I'll replace the following line with .pop() to remove it from the output.

File: directives/ae_output.py [hunk 3]
```diff
@@ -107,10 +108,21 @@ def run(self):
             'max_submissions': self.options.get('submissions', data.get('max_submissions', env.config.ae_default_submissions)),
         })
         data.setdefault('status', self.options.get('status', 'unlisted'))
-        if category in override:
-            data.update(override[category])
-            if 'url' in data:
-                data['url'] = data['url'].format(key=name)
+
+        self.set_url(data, name)
+        self.apply_override(data, category)
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/ae_output.py]: Fixed.

[the remark this answers is not in the document]
Comment (lainets) [reply directives/ae_output.py:112]: You could argue that it is technically submitting things, but I do agree with you. I feel the categories are a bit underused at the moment.

File: directives/questionnaire.py [hunk 4]
```diff
@@ -222,10 +222,14 @@ def run(self):
...
-        if not 'no-override' in self.options and category in override:
-            data.update(override[category])
-            if 'url' in data:
-                data['url'] = data['url'].format(key=name)
+        self.apply_override(data, category)
+        self.set_url(data, name)
+
+        configure_files = {}
+        if data.get("template"):
+            configure_files[data["template"]] = data["template"]
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/questionnaire.py:230]: No, they aren't if this version of a-plus-rst-tools is used. My understanding is that databases doesn't even use a-plus-rst-tools, so it shouldn't be relevant (the Git Manager default(!) settings do copy them, so it should work). The questionnaire directive also doesn't have a `config` option, so its shouldn't even have a container field like a submit directive does.

[the remark this answers is not in the document]
Comment (Mankro) [reply directives/questionnaire.py:230]: Questionnaires can be defined in a yaml file. That yaml file can be used with the `submit` directive and the `config` option. Some courses do that. It's not common to use the `container` field with the grader view type `access.types.stdsync.createForm`, but it is possible and at least the Databases course does that.

The Databases course doesn't use a-plus-rst-tools, that is true.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply README.md:308]: yeah, that's good!

File: directives/questionnaire.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lainets) [reply directives/questionnaire.py:230]: When defining the questionnaire in a yaml file, you need to use the `submit` directive. Then it isn't relevant to the `questionnaire` directive

File: lib/toc_languages.py [hunk 5]
```diff
@@ -179,10 +179,8 @@ def join_exercises(self, key, lang1, c1, lang2, c2):
                 override = self.app.env.config.override.get(c1.get('category'), {})
                 if k in override:
                     c[k] = override[k].format(key=key)
-                elif v != c2.get(k, v):
```
Comment (Mankro) [inline lib/toc_languages.py:182]: Any reason why this `elif` was removed? It's not obvious to me. Is this related to the new gitmanager options?

**Review by Mankro — CHANGES_REQUESTED** [review]
Good! My biggest concern is in the documentation since I don't think it's clear and easy to read for less technical users. I am also concerned about the default URLs, but I don't know how they interact with the new gitmanager. If there is no (default) url set, is the exercise still going to connect to our default MOOC-Grader server grader.cs.aalto.fi (because gitmanager sets the URLs in the build)?

File: aplus_setup.py [hunk 6]
```diff
@@ -50,6 +50,10 @@ def setup(app):
     app.add_config_value('reveal_submission_feedback', None, 'html')
     app.add_config_value('reveal_model_solutions', None, 'html')
     app.add_config_value('enable_autosave', False, 'html')
+    app.add_config_value('unprotected_paths', [], 'html')
+    app.add_config_value('default_exercise_url', None, 'html')
```
[the remark this answers is not in the document]
Comment (lainets) [reply aplus_setup.py:54]: Explained in the new version of Common options for questionnaires and submittable exercises

File: directives/abstract_exercise.py [hunk 7]
```diff
@@ -1,9 +1,25 @@
...
+    Converts a comma-separated list of key:value pairs into a dict.
+    (Directive option conversion function.)
+    """
+    args = [arg.split(":") for arg in argument.split(',')]
+    for arg in args:
+        if len(arg) > 2 or len(arg) == 0:
+            raise SphinxError(
+                "File mapping must be in the format '<mapping 1>,<mapping 2>,...',"
+                " where each <mapping X> is either '<path on grader>:<path in repository>'"
+                " or just '<path>' if no renaming is neccessary.")
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/abstract_exercise.py]: I had though `<>` is a common way to represent placeholders. I guess not. `<mapping 1>` should be fine because `<mapping X> is either ...` is specified right after, right? I understand that it might be confusing for the paths, though. I also added an example to remove any ambiquity.

File: directives/abstract_exercise.py [hunk 8]
```diff
@@ -51,3 +67,79 @@ def set_assistant_permissions(self, data):
 
         if 'allow-assistant-viewing' in self.options:
             data['allow_assistant_viewing'] = str_to_bool(self.options['allow-assistant-viewing'])
+
+class ConfigurableExercise(AbstractExercise):
+    option_spec = {
+        'configure-url': directives.unchanged,
+        'configure-files': file_mapping,
+        'category': directives.unchanged,
+        'no-configure': directives.flag,
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/abstract_exercise.py:77]: `no-override` was only specified for questionnaires before and I didn't want to change the behaviour. I'll add it as you seem to think that `submit` should also support it.

File: directives/abstract_exercise.py [hunk 9]
```diff
@@ -51,3 +67,79 @@ def set_assistant_permissions(self, data):
...
+            raise SphinxError(f"Invalid configure url {url} for {data.get('key', data)}")
+
+        if not parsed.scheme or not parsed.netloc:
+            raise SphinxError(f"Invalid configure url {url} for {data.get('key', data)}")
+
+        extra_files = {}
+        if "configure-files" in data:
+            extra_files = data.pop("configure-files")
+            if not isinstance(extra_files, dict):
+                print(extra_files)
```
[the remark this answers is not in the document]
Comment (lainets) [reply directives/abstract_exercise.py]: Ah, that is a debug print that shouldn't be there

File: lib/toc_languages.py (hunk 5 above)
Comment (lainets) [reply lib/toc_languages.py:182]: I believe this is because the url may contain the exercise key with the language, so they need to be joined. The join_keys will still raise an exception if the URLs are still different after removing the language tag.

File: directives/abstract_exercise.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mankro) [reply directives/abstract_exercise.py:77]: Yes, I think adding it could be useful and at least it can't cause any harm.

**Review by Mankro — APPROVED** [review]
Great, thanks! I think the documentation is much better now. Would you clean up the git commit history and we can merge then?

**Comment by lainets** [comment]
History cleaned
